stream_provider.go

```diff
@@ -254,48 +259,77 @@ func ParseStreamName(fullStreamName string) (string, string) {
 	return fullStreamName, ""
 }
 
-func (r *streamRegistry) register(streamName, dataType string, p *StreamProvider) {
+type StreamDefinitions []*StreamDefinition
+
+func (d StreamDefinitions) StreamNames() string {
+	return d.join(func(sd *StreamDefinition) string {
+		return sd.streamName
+	})
+}
+
+func (d StreamDefinitions) join(f func(*StreamDefinition) string) string {
+	l := make([]string, len(d))
+	for i, v := range d {
+		l[i] = f(v)
+	}
+	return strings.Join(l, ",")
+}
+
+func (d StreamDefinitions) DataTypes() string {
+	return d.join(func(sd *StreamDefinition) string {
+		return sd.dataType
+	})
+}
+
+func (r *streamRegistry) streams() StreamDefinitions {
+	r.Lock()
+	defer r.Unlock()
+	return extractStreamsDefinitions(r)
+}
+
+//To be called only with a lock on the stream registry
+func extractStreamsDefinitions(r *streamRegistry) StreamDefinitions {
+	result := make([]*StreamDefinition, len(r.providers))
```

Comment:
r.Lock()?
Or no multithreading issue possible here?

Reply:
EDIT: mistake, all callers call r.Lock before

Reply:
this method is designed to be called from other methods that already hold a lock on r
and since Go locks are not reentrant...